### PR TITLE
Implement API Routes for any model output type

### PR DESCRIPTION
**Pull request by sunya-ch**
This PR is related to #42.

In this PR, I complete implementation of the model server with API routes supporting for both json weight and archived model under the same format of request. 
 
```mermaid
flowchart LR;
   exporter -- model request in json --> model-server
   model-server -- model weight in json --> exporter
```
example response:
```json
{"All_Weights": 
  {
  "Bias_Weight": 1.0, 
  "Categorical_Variables": {"cpu_architecture": {"Sky Lake": {"weight": 1.0}}}, 
  "Numerical_Variables": {"cpu_cycles": {"mean": 0, "variance": 1.0, "weight": 1.0}}
  }
}
```

```mermaid
flowchart LR;
   exporter -- model request in json --> model-server
   model-server -- model weight per component in json --> exporter
```
example response:
```json 
{
"core": 
  {"All_Weights": 
    {
    "Bias_Weight": 1.0, 
    "Categorical_Variables": {"cpu_architecture": {"Sky Lake": {"weight": 1.0}}}, 
    "Numerical_Variables": {"cpu_cycles": {"mean": 0, "variance": 1.0, "weight": 1.0}}
    }
  },
"dram": 
{"All_Weights": 
  {
  "Bias_Weight": 1.0, 
  "Categorical_Variables": {"cpu_architecture": {"Sky Lake": {"weight": 1.0}}}, 
  "Numerical_Variables": {"cache_miss": {"mean": 0, "variance": 1.0, "weight": 1.0}}
  }
}
```

```mermaid
flowchart LR;
   exporter -- unix.sock --> estimator -- model request in json --> model-server
   model-server -- archived folder of model in zip --> estimator
```
example response: 
```bash
download
├── AbsComponentPower
│   ├── KerasCompFullPipeline.json
│   ├── core
│   │   ├── assets
│   │   ├── keras_metadata.pb
│   │   ├── saved_model.pb
│   │   └── variables
│   │       ├── variables.data-00000-of-00001
│   │       └── variables.index
│   ├── core_transform.pkl
│   ├── dram
│   │   ├── assets
│   │   ├── keras_metadata.pb
│   │   ├── saved_model.pb
│   │   └── variables
│   │       ├── variables.data-00000-of-00001
│   │       └── variables.index
│   ├── dram_transform.pkl
│   └── metadata.json
└── DynPower
    ├── metadata.json
    ├── model.sav
    └── scaler.pkl
```

The model request is mainly composed of list of available metrics, requested output type.
In addition, we can specified specific model name or add the filter string which will be pared by [this filter parser](https://github.com/sunya-ch/kepler-model-server/blob/65f2197288c0df7e54306d95ac411dd92157486c/server/util/loader.py#L18).
```python
class ModelRequest():
    def __init__(self, metrics, output_type, model_name="", filter=""):
        self.metrics = metrics
        self.model_name = model_name
        self.filter = filter
        self.output_type = output_type
```
The model will be automatically selected [here](https://github.com/sunya-ch/kepler-model-server/blob/65f2197288c0df7e54306d95ac411dd92157486c/server/model_server.py#L34).

Check example usage in [model_server_test.py](https://github.com/sunya-ch/kepler-model-server/blob/train-4/tests/model_server_test.py)

[updated according to comments]

### CODE REVIEW

File: server/train/pipeline.py [hunk 1]
```diff
@@ -49,10 +55,17 @@ def update_metadata(self, item):
...
     def train(self, prom_client):
         return NotImplemented
 
-    def get_model_specific_metadata(self):
-        return dict()
+    def archive_model(self):
+        archived_file = get_archived_file(self.group_path, self.model_name) 
+        print("achive model ", archived_file, self.save_path)
+        if os.path.exists(archived_file):
+            os.remove(archived_file)
```
Comment (KaiyiLiu1234) [inline server/train/pipeline.py:69]: I do not think you need to remove the archived file. If it is saved you should be able to just call make_shutil and it should replace it. I will double check.

**Comment by KaiyiLiu1234** [comment]
Also if we are going to be archiving model, should we use SavedModel instead?

**Review by KaiyiLiu1234 — CHANGES_REQUESTED** [review]
I might be misunderstanding something in my previous comments, so feel free to correct me or provide further clarification.

**Comment by sunya-ch** [comment]
> Also if we are going to be archiving model, should we use SavedModel instead?

Sure. You can save to any format in the pipeline train function. We just have to support that in the kepler-estimator.

**Comment by sunya-ch** [comment]
Sorry for updating all changes at the same amend. 
I update the following changes.

- use previously-implemented `return_model_weights` to get weight json. However, combined core and dram to a single return for AbsComponentModelWeight type model.

> ```{"core": {"All_Weights": {"Numerical_Variables": {"cpu_cycles": {"weight": -2.709056854248047, "mean": 6571067392.0, "variance": 6.718171076850876e+18}, "cpu_instr": {"weight": 4.264798641204834, "mean": 3905655552.0, "variance": 4.049098354087428e+18}, "cpu_time": {"weight": 0.20845267176628113, "mean": 168149.25, "variance": 119128200.0}}, "Categorical_Variables": {"cpu_architecture": {"Sandy Bridge": {"weight": 0.5169816613197327}, "Ivy Bridge": {"weight": -0.10627192258834839}, "Haswell": {"weight": 0.10388034582138062}, "Broadwell": {"weight": 20.83293914794922}, "Sky Lake": {"weight": -0.32286643981933594}, "Cascade Lake": {"weight": -0.006997883319854736}, "Coffee Lake": {"weight": 0.5498881936073303}, "Alder Lake": {"weight": -0.18849200010299683}}}, "Bias_Weight": 21.37813377380371}}, "dram": {"All_Weights": {"Numerical_Variables": {"container_memory_working_set_bytes": {"weight": -14.61004638671875, "mean": 217088.0, "variance": 141381599232.0}, "cache_miss": {"weight": -10.457398414611816, "mean": 8926177.0, "variance": 21002180362240.0}}, "Categorical_Variables": {"cpu_architecture": {"Sandy Bridge": {"weight": 0.6605328917503357}, "Ivy Bridge": {"weight": 0.4871968626976013}, "Haswell": {"weight": -0.32830944657325745}, "Broadwell": {"weight": 28.224138259887695}, "Sky Lake": {"weight": -0.15247410535812378}, "Cascade Lake": {"weight": -0.019185900688171387}, "Coffee Lake": {"weight": 0.037583351135253906}, "Alder Lake": {"weight": -0.16830193996429443}}}, "Bias_Weight": 27.499914169311523}}} ```
   
   - similarly update to Sample in AbsModelWeight

- change from .h5 to SavedModel (tf format)
- remove unnecessary remove when archiving model

**Review by KaiyiLiu1234 — APPROVED** [review]
All required changes have been seen.

**Pull request merged**